contrib/exporters/core/extend.go

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright (C) 2019 IBM, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy ofthe License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specificlanguage governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package core
+
+import (
+	"fmt"
+	"time"
+	"encoding/json"
+	"github.com/skydive-project/skydive/api/client"
+	"github.com/skydive-project/skydive/logging"
+	"github.com/spf13/viper"
+	"strings"
+	cache "github.com/pmylund/go-cache"
+)
+
+// extendNone is used for no transform extensions
+type extendNone struct {
+}
+
+func NewExtendNone(cfg *viper.Viper) (interface{}, error) {
+	return &extendNone{}, nil
+}
+
+func (e *extendNone) Extend(in interface{}) interface{} {
+	return in
+}
+
+// extendGremlin is used for transform extensions using gremlin expressions
+// Syntax supported is:
+//	VAR_NAME=<gremlin expression with a single substitution string>
+// For example:
+//	AA_Name=G.V().Has('RoutingTables.Src',%Network.A%).Values('Host')
+// Substitution string must be surrounded with `%`
+// The substitution string refers to fields that exist in the data provided by the particular transform
+// No spaces allowed
+// Use only single quotes in the gremlin expression
+// Need to define `extend` field as a subfield of `transform`
+// Structure in yml file is as follows:
+//   transform:
+//     type: <xxx>
+//     extend:
+//       - VAR_NAME1=<gremlin expression with a single substitution string>
+//       - VAR_NAME2=<gremlin expression with a single substitution string>
+type extendGremlin struct {
+	extendStrings []string
+	gremlinClient *client.GremlinQueryHelper
+	gremlinExprCache *cache.Cache
+}
+
+func NewExtendGremlin(cfg *viper.Viper) (interface{}, error) {
+	gremlinClient := client.NewGremlinQueryHelper(CfgAuthOpts(cfg))
+	extend := cfg.GetStringSlice(CfgRoot + "transform.extend")
+	p := &extendGremlin{
+		extendStrings: extend,
+		gremlinClient: gremlinClient,
+		gremlinExprCache: cache.New(10*time.Minute, 10*time.Minute),
+	}
+	return p, nil
+}
+
+// DeriveValue: use the substitution string to extract the value from the provided map
+func (e *extendGremlin) DeriveValue(S map[string]interface{}, varString string) (string, error) {
```

Comment:
should be a private method deriveValue()

Reply:
This file (core/extend.go) should have been removed. Please ignore code in this file.